dmoj/judge.py

```diff
@@ -175,107 +94,123 @@ def begin_grading(
+                IPC_COMPILE_MESSAGE: self._ipc_compile_message,
+                IPC_GRADING_BEGIN: self._ipc_grading_begin,
+                IPC_GRADING_END: self._ipc_grading_end,
+                IPC_GRADING_ABORTED: self._ipc_grading_aborted,
+                IPC_BATCH_BEGIN: self._ipc_batch_begin,
+                IPC_BATCH_END: self._ipc_batch_end,
+                IPC_RESULT: self._ipc_result,
+                IPC_UNHANDLED_EXCEPTION: self._ipc_unhandled_exception,
+            }
+
+            for ipc_msg in self.current_judge_process.run():
+                type, data = ipc_msg
+
+                handler_func = ipc_handler_dispatch.get(type)
+                if handler_func is None:
+                    raise RuntimeError("unexpected IPC message %s" % str(ipc_msg))
+                handler_func(report, *(data or tuple()))
 
-        # Logs can contain ANSI, and it'll display fine
-        print(message, file=sys.stderr)
+            report(ansi_style('Done grading #ansi[%s](yellow)/#ansi[%s](green|bold).\n' % (problem_id, id)))
+        except:  # noqa: E722, we want to catch everything
+            self.log_internal_error()
+        finally:
+            # In the case that for any reason the judge is still alive, we don't want it to hang around any longer.
+            # We shouldn't ever get here under normal operation barring a race, but might if it died badly.
+            if self.current_judge_process is not None:
+                try:
+                    self.current_judge_process.join(timeout=1)
+                    self.current_judge_process.kill()
+                except OSError:
+                    pass
 
-    def terminate_grading(self):
+            self.current_submission_id = None
+            self.current_judge_process = None
+
+    def _ipc_compile_error(self, report, error_message):
+        report(ansi_style('#ansi[Failed compiling submission!](red|bold)'))
+        report(error_message.rstrip())  # don't print extra newline
+        self.packet_manager.compile_error_packet(error_message)
+
+    def _ipc_compile_message(self, _report, compile_message):
+        self.packet_manager.compile_message_packet(compile_message)
+
+    def _ipc_grading_begin(self, _report, is_pretested):
+        self.packet_manager.begin_grading_packet(is_pretested)
+
+    def _ipc_grading_end(self, _report):
+        self.packet_manager.grading_end_packet()
+
+    def _ipc_result(self, report, batch_number, case_number, result):
+        codes = result.readable_codes()
+
+        # here be cancer
+        is_sc = result.result_flag & Result.SC
+        colored_codes = list(
+            map(lambda x: '#ansi[%s](%s|bold)' % ('--' if x == 'SC' else x, Result.COLORS_BYID[x]), codes)
+        )
+        colored_aux_codes = '{%s}' % ', '.join(colored_codes[1:]) if len(codes) > 1 else ''
+        colored_feedback = '(#ansi[%s](|underline)) ' % utf8text(result.feedback) if result.feedback else ''
+        case_info = (
```

Comment:
This is horrifying. You should definitely do `if not is_sc:`...

Reply:
Done.